cfgov/v1/models/base.py

```diff
@@ -420,16 +420,51 @@ def get_translations(self, inclusive=True, live=True):
         if site:
             pages = pages.in_site(site)
 
-        return pages.order_by("language")
+        language_ordering = {
+            "en": "English",
+            "es": "Español",
+            "zh-Hant": "繁體中文",
+            "vi": "Tiếng Việt",
+            "ko": "한국어",
+            "tl": "Tagalog",
+            "ru": "Pусский",
+            "ar": "العربية",
+            "ht": "Kreyòl Ayisyen",
+        }
+
+        pages = pages.annotate(
+            language_display_order=models.Case(
+                *[
+                    models.When(
+                        language=language,
+                        then=list(language_ordering).index(language),
+                    )
+                    for language in language_ordering
+                ]
+            ),
+            language_display_name=models.Case(
+                *[
+                    models.When(
+                        language=language,
+                        then=models.Value(language_display_name),
+                    )
+                    for language, language_display_name in language_ordering.items()
+                ],
+            ),
+        ).order_by("language_display_order")
```

Comment:
@willbarton I'd love your feedback as to whether this is a clever hack or a terrible idea. I wanted to figure out a way to retain the functionality of get_translations that returns a PageQuerySet.

One option to impact the ordering of the links would be to change the order [in settings](https://github.com/cfpb/consumerfinance.gov/blob/04f3612e53de85d8f7a44404955b303031c2ff03/cfgov/cfgov/settings/base.py#L228-L238); but this feels unexpected and would require us to add sorting everywhere else languages appear. Similarly, to get the custom phrasing we could also change in settings, but then we still need a way to have the English translations appear e.g. when selecting languages in the blog search. Having a hardcoded list feels wrong but seemed the easiest way. I thought about moving the hardcoding somewhere else (ref.py?) but wasn't sure it was worth splitting up the logic. What do you think?

Reply:
My initial discomfort is that, I don't like duplicating the language list from settings on lines to 423-432. And in the [eyebrow template as a whole](https://github.com/cfpb/consumerfinance.gov/blob/ec96959b87e05c66af5dc95377c1f03549eeae4d/cfgov/v1/jinja2/v1/includes/molecules/global-eyebrow.html#L26-L75).

I think if we're going to define a sort order, it should be in settings, and used everywhere (eyebrow, here, anywhere else we want translation links). Because we already have an ordered tuple in settings (and, AFAIR, we can rely on tuples order being preserved), couldn't `language_ordering` being something like:

```python
language_ordering = (lang_code for lang_code, lang_name in settings.LANGUAGES)
```

More than anything, I think that duplication of the list of languages bothers me — it means we'd have to change multiple places where we define what languages we support, not all of them obvious (settings for what the `language` field lists, here for what we offer the public on pages with translations, and the eyebrow template for the eyebrow). I have another thought on this, but it's only tangentially related to this PR, so I'll make it in a separate comment.

As far as the annotation goes though? I think it makes sense, and is kind of the use-case for it. I think I'd just want to have it well-commented, so that anyone who encounters it for debugging understands the intention. 

My only other thought was, what would migration looks like from sorting like this to Wagtail i18n's `TranslatableMixin`, and it seems to rely on the [`Locale` model's ordering](https://github.com/wagtail/wagtail/blob/main/wagtail/models/i18n.py#L51-L54)... which, I think, would mean this overall approach of annotating the queryset should still work, with some modification.

I also wonder if it would be overkill to (maybe eventually) move that sorting logic off into a `PageQuerySet` subclass so it could be more isolated and testable. But that's probably only worth doing after we migrate to Wagtail i18n and see what accounting for the `locale` foreign key's sorting.

---

> I thought about moving the hardcoding somewhere else (ref.py?)

![No.](https://user-images.githubusercontent.com/10562538/214064604-58a44945-49f8-4745-a117-570c82a340e0.gif)

Reply:
Once this PR is merged, I'll go about consolidating the translations and language names and all that in #7392.